src/Storages/Kafka/StorageKafka.cpp

```diff
@@ -268,69 +268,103 @@ namespace
-            const String kafka_topic_path = config_prefix + "." + tag;
-            const String kafpa_topic_name_path = kafka_topic_path + "." + CONFIG_NAME_TAG;
-
-            const String topic_name = config.getString(kafpa_topic_name_path);
-            if (topic_name == topic)
+            if (tag == CONFIG_KAFKA_TOPIC_TAG) /// multiple occurrences given as "kafka_topic", "kafka_topic[1]", etc.
             {
-                /// Found it! Now read the per-topic configuration into cppkafka.
-                Poco::Util::AbstractConfiguration::Keys inner_tags;
-                config.keys(kafka_topic_path, inner_tags);
-                for (const auto & inner_tag : inner_tags)
+                // Update consumer topic-specific configuration (new syntax). Example with topics "football" and "baseball":
+                //     <kafka>
+                //         <kafka_topic>
+                //             <name>football</name>
+                //             <retry_backoff_ms>250</retry_backoff_ms>
+                //             <fetch_min_bytes>5000</fetch_min_bytes>
+                //         </kafka_topic>
+                //         <kafka_topic>
+                //             <name>baseball</name>
+                //             <retry_backoff_ms>300</retry_backoff_ms>
+                //             <fetch_min_bytes>2000</fetch_min_bytes>
+                //         </kafka_topic>
+                //     </kafka>
+                // Advantages: The period restriction no longer applies (e.g. <name>sports.football</name> will work), everything
+                // Kafka-related is below <kafka>.
+                for (const auto & topic : topics)
                 {
-                    if (inner_tag == CONFIG_NAME_TAG)
-                        continue; // ignore <name>
+                    /// Read topic name between <name>...</name>
+                    const String kafka_topic_path = config_prefix + "." + tag;
+                    const String kafpa_topic_name_path = kafka_topic_path + "." + CONFIG_NAME_TAG;
 
-                    /// "log_level" has valid underscore, the remaining librdkafka setting use dot.separated.format which isn't acceptable for XML.
-                    /// See https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
-                    const String setting_path = kafka_topic_path + "." + inner_tag;
-                    const String setting_value = config.getString(setting_path);
+                    const String topic_name = config.getString(kafpa_topic_name_path);
 
-                    const String setting_name_in_kafka_config = (inner_tag == "log_level") ? inner_tag : boost::replace_all_copy(inner_tag, "_", ".");
-                    kafka_config.set(setting_name_in_kafka_config, setting_value);
+                    if (topic_name != topic)
+                        continue;
+                    loadTopicConfig(kafka_config, config, kafka_topic_path);
```

Comment:
I was thinking that separate consumer/producer settings should support per-topic syntax as well, and both of them, this one and simpler.

Reply:
Do you mean `<kafka_TOPIC_NAME>`? I thought that it is a part of the legacy... (regarding code comments)

Maybe it is better to reject `<kafka_TOPIC_NAME>` at all? Simpler code / less config variations.

Reply:
>Do you mean <kafka_TOPIC_NAME>? I thought that it is a part of the legacy... (regarding code comments)

Yes, it is deprecated syntax, but maybe it still make sense to support it, since it is not a lot of code, but easier syntax (usually you don't have special symbols in the topic names and you are allowed to use easier syntax instead of more "advanced " one).

But I don't have strong opinion here, I'm OK to leave it as-is.

Reply:
Moved the legacy topic parsing to a separate function (will be easier to deprecate in the future) and allowed to use it (old schema) inside `<consumer>/<producer>` tags.